### ISSUE TITLE
IcingaDB Services crashes while syncing to MariaDB

### ISSUE THREAD

**Issue by Stefan-Hoesl** [Icinga/icingadb#442]
## Describe the bug

We have installed Icinga2 with 2 Master and 2 Satellites.
On both Master IcingaDB ist installed and configured.

After Icinga2 start the IcingaDB Service begins to write into the DB
and chrashes after a short time with the messages shown below:
On Master01:
```
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[361738]: Starting config sync
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[361738]: Starting overdue sync
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[361738]: Starting initial state sync
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[361738]: config-sync: Updating 1751 items of type host state
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[361738]: config-sync: Inserting 1774 items of type checkcommand argument
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[361738]: config-sync: Aborted config sync after 171.128965ms
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[361738]: config-sync: Aborted initial state sync after 171.95646ms
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[361738]: json: cannot unmarshal number 2.2 into Go value of type int64
                                                             can't unmarshal JSON into *int64
                                                             github.com/icinga/icingadb/internal.UnmarshalJSON
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/internal/internal.go:47
                                                             github.com/icinga/icingadb/pkg/types.(*Int).UnmarshalJSON
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/pkg/types/int.go:52
                                                             encoding/json.(*decodeState).literalStore
                                                                     /usr/local/go/src/encoding/json/decode.go:860
                                                             encoding/json.(*decodeState).value
                                                                     /usr/local/go/src/encoding/json/decode.go:384
                                                             encoding/json.(*decodeState).object
                                                                     /usr/local/go/src/encoding/json/decode.go:765
                                                             encoding/json.(*decodeState).value
                                                                     /usr/local/go/src/encoding/json/decode.go:370
                                                             encoding/json.(*decodeState).unmarshal
                                                                     /usr/local/go/src/encoding/json/decode.go:180
                                                             encoding/json.Unmarshal
                                                                     /usr/local/go/src/encoding/json/decode.go:107
                                                             github.com/icinga/icingadb/internal.UnmarshalJSON
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/internal/internal.go:47
                                                             github.com/icinga/icingadb/pkg/icingaredis.CreateEntities.func1.1
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/pkg/icingaredis/utils.go:56
                                                             golang.org/x/sync/errgroup.(*Group).Go.func1
                                                                     /home/build/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57
                                                             runtime.goexit
                                                                     /usr/local/go/src/runtime/asm_amd64.s:1371
                                                             can't unmarshal JSON into *v1.CheckcommandArgument
                                                             github.com/icinga/icingadb/internal.UnmarshalJSON
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/internal/internal.go:47
                                                             github.com/icinga/icingadb/pkg/icingaredis.CreateEntities.func1.1
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/pkg/icingaredis/utils.go:56
                                                             golang.org/x/sync/errgroup.(*Group).Go.func1
                                                                     /home/build/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57
                                                             runtime.goexit
                                                                     /usr/local/go/src/runtime/asm_amd64.s:1371
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx systemd[1]: icingadb.service: Main process exited, code=exited, status=1/FAILURE
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx systemd[1]: icingadb.service: Failed with result 'exit-code'.
```
On Master02:
```
Feb 17 14:25:28 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[213366]: config-sync: Inserting 1774 items of type checkcommand argument
Feb 17 14:25:28 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[213366]: config-sync: Inserting 7054 items of type hostgroup member
Feb 17 14:25:29 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[213366]: config-sync: Inserting 20782 items of type host customvar
Feb 17 14:25:29 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[213366]: config-sync: Aborted initial state sync after 560.725858ms
Feb 17 14:25:29 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[213366]: config-sync: Inserting 20300 items of type notification usergroup
Feb 17 14:25:29 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[213366]: config-sync: Inserting 24538 items of type notification customvar
Feb 17 14:25:29 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[213366]: config-sync: Aborted config sync after 563.542894ms
Feb 17 14:25:29 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[213366]: json: cannot unmarshal number 2.1 into Go value of type int64
                                                             can't unmarshal JSON into *int64
                                                             github.com/icinga/icingadb/internal.UnmarshalJSON
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/internal/internal.go:47
                                                             github.com/icinga/icingadb/pkg/types.(*Int).UnmarshalJSON
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/pkg/types/int.go:52
                                                             encoding/json.(*decodeState).literalStore
                                                                     /usr/local/go/src/encoding/json/decode.go:860
                                                             encoding/json.(*decodeState).value
                                                                     /usr/local/go/src/encoding/json/decode.go:384
                                                             encoding/json.(*decodeState).object
                                                                     /usr/local/go/src/encoding/json/decode.go:765
                                                             encoding/json.(*decodeState).value
                                                                     /usr/local/go/src/encoding/json/decode.go:370
                                                             encoding/json.(*decodeState).unmarshal
                                                                     /usr/local/go/src/encoding/json/decode.go:180
                                                             encoding/json.Unmarshal
                                                                     /usr/local/go/src/encoding/json/decode.go:107
                                                             github.com/icinga/icingadb/internal.UnmarshalJSON
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/internal/internal.go:47
                                                             github.com/icinga/icingadb/pkg/icingaredis.CreateEntities.func1.1
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/pkg/icingaredis/utils.go:56
                                                             golang.org/x/sync/errgroup.(*Group).Go.func1
                                                                     /home/build/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57
                                                             runtime.goexit
                                                                     /usr/local/go/src/runtime/asm_amd64.s:1371
                                                             can't unmarshal JSON into *v1.CheckcommandArgument
                                                             github.com/icinga/icingadb/internal.UnmarshalJSON
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/internal/internal.go:47
                                                             github.com/icinga/icingadb/pkg/icingaredis.CreateEntities.func1.1
                                                                     /builds/packaging/rpm-icingadb/build/BUILD/icingadb-1.0.0-rc2/pkg/icingaredis/utils.go:56
                                                             golang.org/x/sync/errgroup.(*Group).Go.func1
                                                                     /home/build/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57
                                                             runtime.goexit
                                                                     /usr/local/go/src/runtime/asm_amd64.s:1371
Feb 17 14:25:29 xxxxxxxxxxxxxxxxxxxxxxxxxx systemd[1]: icingadb.service: Main process exited, code=exited, status=1/FAILURE
Feb 17 14:25:29 xxxxxxxxxxxxxxxxxxxxxxxxxx systemd[1]: icingadb.service: Failed with result 'exit-code'.
```
The difference in output is:
```
Feb 17 14:21:48 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[361738]: json: cannot unmarshal number 2.2 into Go value of type int64
Feb 17 14:25:29 xxxxxxxxxxxxxxxxxxxxxxxxxx icingadb[213366]: json: cannot unmarshal number 2.1 into Go value of type int64
```

IcingaDB was successfully setup on a single node installation.

## Your Environment

Include as many relevant details about the environment you experienced the problem in

* Icinga DB version: 1.0.0-rc2
* Icinga 2 version: 2.13.2-1
* Operating System and version: RHEL8

**Comment by Stefan-Hoesl** [comment]
Are there any comments on this issue

**Comment by lippserd** [comment]
Hi,

This has been fixed in Icinga 2. I think u are using floats for the order of command arguments. If you switch to ints, there should be no more errors. You may also use Icinga snapshot packages.

All the best,
Eric

**Comment by Stefan-Hoesl** [comment]
Found the order definition and set INT values.
Now the next error arises:

![IcingaDB-Crash3](https://user-images.githubusercontent.com/99953967/158365177-64296883-39b5-4757-83f0-f090f8e03a30.png)

**Comment by Stefan-Hoesl** [comment]
The Problem was caused by Galera-Cluster and HAProxy.

IcingaDB is now up and running.

**Comment by jwhite-ac** [comment]
> The Problem was caused by Galera-Cluster and HAProxy.
> 
> IcingaDB is now up and running.

Could you elaborate what the issue was with Galera/HAproxy? 

I'm running a similar setup (2 masters, DB on Galera cluster fronted by HAproxy), and am seeing regular deadlocks for icingadb, which crash the process (instead of the process retrying, so I've had to set a systemd drop-in for icingadb to set `Restart: always` which stops icingadb from crashing and staying stopped, but I'm still seeing a lot of deadlocks so wonder if the issue you had is the cause of what I'm seeing (or related to it).

**Comment by oxzi** [comment]
@jwhite-ac: This response might be kind of late, but you might wanna take a look at #577.
As you're already using HAProxy, setting `balance first` or `balance source` could be a quick fix.

**Comment by jwhite-ac** [comment]
> @jwhite-ac: This response might be kind of late, but you might wanna take a look at #577. As you're already using HAProxy, setting `balance first` or `balance source` could be a quick fix.

At the time I was investigating an issue that was causing the galera cluster to lock-up after deadlocks very slowly cascaded and took out each node, but that has since "resolved itself" (I'm not sure if it was a bug regression as I'd seen it on another cluster about a year prior, but much worse as it was a busier cluster - no Icinga2 DBs there though). When I commented here I was chasing the lead of all the deadlocks I was seeing for Icinga2, thinking it might be the (or part of the) problem, but it turns out it was a red herring and the Icinga2 deadlocks, while not great, are handled without issue by MariaDB/Galera.

I've not had any issues with icingadb since adding the systemd drop-in to set the process to always restart on fail forever, and have made more configuration adjustments as I've learned more about how the multi-master set up is meant to be configured for proper live-live availability.

I can see how using balance first or balance source (probably the latter rather than the former) would likely help with reducing the deadlocks, since you'd be writing to the same node (more or less) each time (or the same two nodes in a dual master setup). I may give that a got at some point and see what effect it has, but for the last few months (or more) the deadlocks have not been a problem and icingadb itself comes back from any failures (if they're still happening).
